Add RoPE (Rotary Positional Embedding) to PyTorch core

sladyn98 opened the pull request:
This PR adds RoPE (Rotary Positional Embedding) to PyTorch core to centralize the implementation and reduce code duplication across PyTorch repositories.

Fixes #149534

vadimkantorov commented:
Also, I guess there are other existing impls in HF, unsloth etc:
- https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/rope_embedding.py

with some differing details. I wonder, if they would be able to replace their impls with PyTorch built-in one. And what are the tiny details (e.g. upcasting/downcasting/compute dtype) which are important for being able to replace existing impls with the new built-in one...

Commit:
Fix issue #149534: Addind RoPE to pytorch core

Implement RoPE (Rotary Positional Embedding) in PyTorch core to eliminate
code duplication across PyTorch repositories.

- Add functional API: rotary_position_embedding_freqs() and rotary_position_embedding()
- Add module API: RotaryPositionalEmbedding class in nn.modules.transformer
- Include comprehensive error handling and gradient support
- Add detailed documentation and usage examples
- Support arbitrary batch dimensions and flexible device/dtype
